Tutorial: How to set up one time feature flags (#5955)
PostHog/posthog.com: contents/tutorials/one-time-feature-flags.md

---
title: How to set up one-time feature flags
date: 2023-05-12
author: ["ian-vanagas"]
showTitle: true
sidebar: Docs
featuredImage: ../images/tutorials/banners/tutorial-4.png
tags: ['feature flags']
---

Sometimes you want to show users a component or some content only once. You use a field in their user model or store it locally, but this can get messy fast. It also might prevent you from changing it remotely. A better way to do this is a feature flag that changes once a user completes what you want.

> In this tutorial, we show how to set up a one-time feature flag by building a basic Express.js API that provides one response on the first request, and a different one on subsequence requests.

Reviewer: Maybe?
```suggestion
In this tutorial, we show how to set up a one-time feature flag using PostHog by building a basic Express.js API that provides one response on the first request, and a different one on subsequence requests.
```

## Creating our Express app

To create our basic Express app, first, install Express.

```bash
npm i express
```

Next, create a basic Express app that returns "Hello World!" on requests to the basic route.

```js
// server.js
const express = require('express');
const app = express();
const port = 3000;

app.get('/', (req, res) => {
  res.send('Hello World!')
})

app.listen(port, () => {
  console.log(`Example app listening on port ${port}`)
})
```

After setting this up, you can run the server.

```bash
node server.js
```

This is the basic app set up to showcase one-time feature flags.

## Setting up PostHog

Next, install PostHog through the [Node SDK](/docs/libraries/node) as Express is a Node framework.

```bash
npm i posthog-node
```

> Once done require `posthog-node` in our Express app and set up the PostHog client with our project API key and instance address. You can find these in the getting started flow or in your [project settings](https://app.posthog.com/project/settings)

Reviewer: ```suggestion
Once done, add `posthog-node` to your Express app and set up the PostHog client with our project API key and instance address. You can find these in the getting started flow or in your [project settings](https://app.posthog.com/project/settings)
```

```js
// server.js
//...
const { PostHog } = require('posthog-node');
const client = new PostHog(
    '<ph_project_api_key>',
    { host: '<ph_instance_address>'}
)
//...
```

> **Optional:** if you want your flags to evaluate as fast as possible, you can set up local evaluation by including a personal API key in your client initialization. You can create a personal API key in your [account settings](https://app.posthog.com/me/settings).

### Capture the first time event

Once PostHog is set up, create a custom event that triggers when a user makes a request to the base route. Make sure to both:

> 1. Include a user ID in the request. I did so by including an ID param in the URL like `http://localhost:3000/?id=ian` then handling with `req.query.id`.

Reviewer: ```suggestion
1. Include a user ID in the request. I did so by including an ID param in the URL like `http://localhost:3000/?id=ian`, and then handling with `req.query.id`.
```
